README.md
![GitHub Actions CI status](https://github.com/GrafeasGroup/tor/workflows/Automated%20Testing/badge.svg)

# Transcribers of Reddit

This is the source code for the bot moderating and managing several parts of the subreddit
[/r/TranscribersOfReddit](https://reddit.com/r/TranscribersOfReddit) ("ToR"), a community dedicated to transcribing images, audio, and video. It acts under the username "[/u/TranscribersOfReddit](https://reddit.com/u/TranscribersOfReddit)".

Among other things, this bot handles:

- Posting transcription requests to /r/TranscribersOfReddit as relevant content shows up on partner subreddits
- Responding to transcribers claiming and marking requests as complete
- Augmenting the score of a transcriber upon successful completion of a transcription

## Requirements

- Redis (tracking completed posts and queue system)
- Reddit API keys

> **NOTE:**
>
> This code is not complete. The praw.ini file is required to run the bots and
> contains such information as the useragents and various secrets. It is built
> for Python 3.6.

## Installation

### From release

Given a release in <https://github.com/GrafeasGroup/tor/releases> (perhaps the [latest](https://github.com/GrafeasGroup/tor/releases/latest)), download the attached `.whl` file for your platform/architecture and `pip install` it directly like so:

```sh
$ pip install ./path/to/tor-3.6.1-py3-any-none.whl
```

### From source

Make sure you have an [up-to-date copy of pip installed](https://pip.pypa.io/en/stable/installing/), the latest version of [Poetry](https://www.python-poetry.org/), and Python 3.6.

```sh
$ git clone https://github.com/GrafeasGroup/tor.git tor
$ cd tor/
$ poetry install
```

## Big Picture

Triggered flow (via bot inbox):

- If username mention in a comment:
  - Grab id of mentioned post's parent
  - Make new post on ToR
- If reply to comment (e.g., `claim` or `done`):
  - `claim` - Assigns transcription ownership of the post to the user who first commented this way
  - `done` - Checks for transcription and adjusts flair according to reward system

Monitoring daemon (via subreddit's /new feed):

- For each subreddit that has opted in:
  - Search for audio, video, and image content:
    - Check against whitelist of domain filters
    - Post url to the content back to /r/TranscribersOfReddit

## Build

To build the package from source, start in the base of the repository and run:

```sh
$ poetry build
```

When building is complete, upload everything in the `dist/` directory that was just created as part of the GitHub release.

## Usage

```sh
$ tor-moderator
# => [daemon mode + logging]
```

## Contributing

See [`CONTRIBUTING.md`](/CONTRIBUTING.md) for more.

## Sponsors

<a href="https://bugsnag.com"><img src="https://raw.githubusercontent.com/GrafeasGroup/tor/master/images/bugsnag_logo_navy.png" alt="Bugsnag logo" width=130></a>

For providing free error handling through [their open source program](https://www.bugsnag.com/open-source/)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [UNRELEASED]

- Track timestamp of last reddit post processed before sending to mod chat (credit: @crhopkins)
- Add missing backslash in unescaped heading string (credit: @MurdoMaclachlan)

## [4.2.4] - 2021-04-05

- Odd error with PRAW necessitates a core library upgrade

## [4.2.3] - 2021-03-24

- Hotfix for private subreddits

## [4.2.2] - 2021-01-17

- Fixes bug introduced by Reddit where bot cannot post on other subs, thus cross-posting capability was removed

## [4.2.1] - 2021-01-06

- Fixes args used in CLI invocation, making testing if it is installed correctly work more smoothly

## [4.2.0] - 2020-12-26

- Includes flair for "topaz" and "jade" levels

## [4.1.0] - 2020-03-03

- Update logging to note the current transcription number and minor testing config changes

## [4.0.0] - 2020-02-29

- Minor typos in bot messages
- Removes heartbeat server
- Adds type annotations to method signatures
- Removes a metric crap-ton of dead code
- Clarifies the logic of what the bot is doing and why, using more semantic names of methods
- Tightens some overly broad interfaces (Only default value of a keyword arg is used? Buh-bye keyword arg, you're now permanently set that way!)
- Uses flake8, mypy, and pytest as part of CI execution
- Replaces Travis CI with GitHub Actions

## [3.12.2] - 2020-02-29

- Fixes bug in build system (Poetry) where command is not generated

## [3.12.1] - 2020-02-29

- Skip mod flags that (somehow) come back as `null`, because obviously human intervention is needed (credit @thelonelyghost)
- Add link to tutorial for first transcriptions (credit @Pf441)
- Mentions donation option in (successful) response to `done` (credit @thelonelyghost)

## [3.12.0] - 2019-12-12

- Replaces `setup.py` with Poetry tooling for development and packaging ease
- Fixes no-author condition if the mods message the bot
- Adds `cancel` as an alias for `unclaim` for user directives to the bot
- Makes inbox triage less English-language dependent
- Offloads more work to Redis instead of Python

## [3.11.2] - 2019-09-14

- Adds missing `package_data` manifest to `setup.py` so it will actually be included when packaging the module (credit @thelonelyghost)

## [3.11.1] - 2019-09-14

- Fixes typo in capitalization of one of the user-facing messages
- Fixes flair colors to level up at (e.g.) 50, not 51 (#96)
- Extend check for transcription in the right place to handle New Reddit editor (credit @itsthejoker)
- Fixes packaging of python module to include data files and up-to-date methods (credit @thelonelyghost)

## [3.11.0] - 2019-06-17

- Add replies to DMs (credit: @davidarchibald)

## [3.10.1] - 2019-06-17

- Fixes history check on users who have submissions stickied to their profile (credit @arfie)
- Fixes youtube auto-transcription attempts erroring out when unable to detect the video id (credit @thelonelyghost)
- Adds clearer indicator for post type to rules comment (credit @itsthejoker)

## [3.10.0] - 2019-06-16

- Fixes [#102](https://github.com/GrafeasGroup/tor/issues/102), adding better guidance for first-time transcribers
- Allows CLI flags like `--help`, `--version`, `--debug`, and `--noop` on `tor-moderator`

## [3.9.0] - 2019-06-16

- Removes remaining references to RocketChat
- Fixes typo in YAML dict name
- Adds more context to Bugsnag events
- Corrects version reported to Bugsnag to be `tor`'s, not `tor.core`'s

## [3.8.0] - 2019-06-16

- Removes Sentry client and dependency
- Replaces strings.py references with YAML

## [3.7.0] - 2019-06-16

- Adds No Operation (NOOP) mode
- Merges in `tor_core` as `tor.core` (Fixes [#150](https://github.com/GrafeasGroup/tor/issues/150))
- Uses `tox` as the default testing mechanism (easier CI and enforces virtualenv sandbox testing)

## [3.6.2] - 2019-06-15

- FIX: `Reddit.user.me` is a method not a property, resulting in stack traces on every attempt to run the bot

## [3.6.1] - 2019-06-07

- HOTFIX release:
  - Makes the (protected) bot names an environment variable we can override in case usernames need to suddenly change
  - Add reference to the currently running bot's username, as determined by Reddit's API

## [3.6.0] - 2018-12-02

- Add handling for `unclaim` comments (credit: @itsthejoker)

## [3.5.0] - 2018-08-19

- Add check to verify the transcription goes to the right place (credit: @itsthejoker)
- Allow dibs to claim a post (credit: @davidarchibald)
- Update Slack messages to include direct links (credit: @pejmanpoh)
- Remove check for incomplete posts (credit: @itsthejoker)
- Add threaded model to update the process of checking for new content (credit: @itsthejoker)
- Move volunteer notifications to their own channel (credit: @itsthejoker)
- Update detection of flair (credit: @itsthejoker)
- Add system to use history of volunteer to validate posts if comment is autoremoved (credit: @itsthejoker)

## [3.4.0] - 2017-12-11

- Send inbox messages to Slack if they don't match any of the commands (credit: @arfie)
- Send message to Slack on phrases that may need mod intervention (credit: @thelonelyghost)
- Allow blacklisting of people (credit: @perryprog)
- Add new command system (credit: @perryprog)
- Use the same colors for subreddit flairs as on Discord, depending on number of completed transcriptions (credit: @itsthejoker)
- Send message to Slack when flairing a post as Meta (credit: @arfie)
- Handles common typo `deno` as `done` (credit: @itsthejoker)

## [3.3.0] - 2017-11-22

- Enable alternate validation method to get around spam filter nuking posts (credit: @itsthejoker)

## [3.2.0] - 2017-11-07

- Update inbox handler to increase handling speed
- Update inbox handler to increase legibility and modularity
- Guard every comment reply in case it gets deleted

tor/cli/main.py
import atexit
import logging
import os
import pathlib
import sys
import time

import beeline
import click
from click.core import Context
from dotenv import load_dotenv
from praw import Reddit
from shiv.bootstrap import current_zipfile  # type: ignore

# The `import tor` lines is necessary because `tor.__SELF_NAME__` is
# set here. Reason: https://gist.github.com/TheLonelyGhost/9dbe810c42d8f2edcf3388a8b19519e1
import tor
from tor import __version__
from tor.core.config import config
from tor.core.helpers import run_until_dead
from tor.core.inbox import check_inbox
from tor.core.initialize import initialize
from tor.helpers.flair import set_meta_flair_on_other_posts
from tor.helpers.threaded_worker import threaded_check_submissions

##############################
NOOP_MODE = bool(os.getenv("NOOP_MODE", ""))
DEBUG_MODE = bool(os.getenv("DEBUG_MODE", ""))
##############################

# Patreon Dedications:
#
# Through our work supported by Grafeas Group Ltd., some people see
# fit to help us keep the lights on. Of those helpful souls, even
# fewer gift $50 or more and they have earned their place in the
# following list. For reference, the link can be found here:
# https://www.patreon.com/grafeasgroup
# List is in date order.
#
# Jake L (2017-11-17)
# Michael W (2017-11-27)
# Luke Abby (2019-07-02)
# David G (2021-06-14)
# Jason H (2021-07-01)

# Musical Dedications:
#
# This program is dedicated to the below artists; their music has
# served as the soundtrack for the continued development of u/ToR.
# List is in alphabetical order. Anyone who contributes to this
# codebase is invited to add their tunes!
#
# Alec Benjamin
# Alison Wonderland
# Apashe
# Aramanthe
# Betty Who
# blink-182
# Braxton Burks
# Caravan Palace
# Daft Punk
# David Bowie
# DIAMANTE
# Dorothy
# Flyleaf
# Halestorm
# Hiromi
# Girl Talk
# Green Day
# Icon for Hire
# Inverness
# John Williams
# K-391
# Lady Gaga
# Neon Hitch
# Queen
# Rage Against the Machine
# Rita Ora
# The Beatles
# The Killers
# The Pretty Reckless
# Two Door Cinema Club
#
#
# Streams:
# https://www.youtube.com/watch?v=hX3j0sQ7ot8  # he's dead, Jim

log = logging.getLogger()

with current_zipfile() as archive:
    dotenv_path: str | None
    if archive:
        # if archive is none, we're not in the zipfile and are probably
        # in development mode right now.
        dotenv_path = str(pathlib.Path(archive.filename).parent / ".env")
    else:
        dotenv_path = None
load_dotenv(dotenv_path=dotenv_path)


def run_noop(cfg):
    pass


@beeline.traced(name="run")
def run(cfg):
    """
    Primary routine.

    :param cfg: Global config dict, supplied by tor_core.
    :return: None.
    """
    check_inbox(cfg)

    threaded_check_submissions(cfg)

    set_meta_flair_on_other_posts(cfg)

    if cfg.debug_mode:
        time.sleep(15)


@click.group(
    context_settings=dict(help_option_names=["-h", "--help", "--halp"]),
    invoke_without_command=True,
)
@click.pass_context
@click.option(
    "-d",
    "--debug",
    "debug",
    is_flag=True,
    default=DEBUG_MODE,
    help="Puts bot in dev-mode using non-prod credentials",
)
@click.option(
    "-n",
    "--noop",
    "noop",
    is_flag=True,
    default=NOOP_MODE,
    help="Just run the daemon, but take no action (helpful for testing infrastructure changes)",
)
@click.version_option(version=__version__, prog_name=tor.__SELF_NAME__)
def main(ctx: Context, debug: bool, noop: bool):
    """Run ToR."""
    if ctx.invoked_subcommand:
        # If we asked for a specific command, don't run the bot. Instead, pass control
        # directly to the subcommand.
        return

    logging.basicConfig(
        level=logging.INFO,
        format="%(levelname)s | %(funcName)s | %(message)s",
        datefmt="%Y-%m-%dT%H:%M:%S",
    )

    honeycomb_key = os.getenv("HONEYCOMB_KEY", "")

    # if no api key, do not send data and instead print what would be sent to stderr
    # if we have a key, pass data to honeycomb.io
    args = {
        "writekey": honeycomb_key,
        "dataset": "transcribersofreddit",
        "service_name": "tor_moderator",
        "debug": True if len(honeycomb_key) == 0 else False,
        "sample_rate": 10,
    }
    beeline.init(**args)
    atexit.register(beeline.close)

    config.debug_mode = debug

    if config.debug_mode:
        bot_name = "debug"
    else:
        bot_name = os.environ.get("BOT_NAME", "bot")

    log.info(f"Connecting to Reddit as {bot_name}.")
    config.r = Reddit(
        user_agent=bot_name,
        client_id=os.environ.get("REDDIT_CLIENT_ID", ""),
        client_secret=os.environ.get("REDDIT_SECRET", ""),
        username=os.environ.get("REDDIT_USERNAME", ""),
        password=os.environ.get("REDDIT_PASSWORD", ""),
    )
    initialize(config)
    config.perform_header_check = True
    log.info("Bot built and initialized")

    tor.__SELF_NAME__ = config.r.user.me().name
    if tor.__SELF_NAME__ not in tor.__BOT_NAMES__:
        tor.__BOT_NAMES__.append(tor.__SELF_NAME__)

    if noop:
        run_until_dead(run_noop)
    else:
        run_until_dead(run)


@main.command()
@click.option(
    "-v",
    "--verbose",
    is_flag=True,
    default=False,
    help="Show Pytest output instead of running quietly.",
)
def selfcheck(verbose: bool) -> None:
    """
    Verify the binary passes all tests internally.

    Add any other self-check related code here.
    """
    import pytest

    import tor.test

    # -x is 'exit immediately if a test fails'
    # We need to get the path because the file is actually inside the extracted
    # environment maintained by shiv, not physically inside the archive at the
    # time of running.
    args = ["-x", str(pathlib.Path(tor.test.__file__).parent)]
    if not verbose:
        args.append("-qq")
    # pytest will return an exit code that we can check on the command line
    sys.exit(pytest.main(args))


BANNER = r"""
___________   __________
\__    ___/___\______   \
  |    | /  _ \|       _/
  |    |(  <_> )    |   \
  |____| \____/|____|_  /
                      \/
"""


@main.command()
def shell() -> None:
    """Create a Python REPL inside the environment."""
    import code

    code.interact(local=globals(), banner=BANNER)


if __name__ == "__main__":
    main()
